test/http_tiles.cc

```diff
@@ -0,0 +1,224 @@
+#include "test.h"
+
+#include <prime_server/http_protocol.hpp>
+#include <prime_server/http_util.hpp>
+#include <prime_server/prime_server.hpp>
+
+#include <boost/property_tree/ptree.hpp>
+
+#include <chrono>
+#include <csignal>
+#include <functional>
+#include <stdexcept>
+#include <thread>
+
+#include "baldr/compression_utils.h"
+#include "baldr/rapidjson_utils.h"
+#include "filesystem.h"
+#include "tyr/actor.h"
+
+using namespace prime_server;
+using namespace valhalla;
+
+std::string gzip(std::string& uncompressed) {
+  auto deflate_src = [&uncompressed](z_stream& s) {
+    s.next_in = static_cast<Byte*>(static_cast<void*>(&uncompressed[0]));
+    s.avail_in = static_cast<unsigned int>(uncompressed.size() * sizeof(std::string::value_type));
+    return Z_FINISH;
+  };
+
+  std::string compressed;
+  auto deflate_dst = [&compressed](z_stream& s) {
+    // if the whole buffer wasn't used we are done
+    auto size = compressed.size();
+    if (s.total_out < size)
+      compressed.resize(s.total_out);
+    // we need more space
+    else {
+      // set the pointer to the next spot
+      compressed.resize(size + 16);
+      s.next_out = static_cast<Byte*>(static_cast<void*>(&compressed[0] + size));
+      s.avail_out = 16;
+    }
+  };
+
+  if (!baldr::deflate(deflate_src, deflate_dst))
+    throw std::logic_error("Can't write gzipped string");
+
+  return compressed;
+}
+
+worker_t::result_t
+disk_work(const std::list<zmq::message_t>& job, void* request_info, worker_t::interrupt_function_t&) {
+  worker_t::result_t result{false};
+  auto* info = static_cast<http_request_info_t*>(request_info);
+  try {
+    // parse request
+    const auto request =
+        http_request_t::from_string(static_cast<const char*>(job.front().data()), job.front().size());
+
+    // ends with gz
+    auto path = request.path;
+    auto gz = path.find(".gz");
+    if (gz == path.size() - 3)
+      path = path.substr(0, gz);
+    else
+      gz = 0;
+
+    // load the file and gzip it if we have to
+    std::string full_path =
+        "test/data/utrecht_tiles" + (filesystem::path::preferred_separator + path);
+    std::fstream input(full_path, std::ios::in | std::ios::binary);
+    if (input) {
+      std::string buffer((std::istreambuf_iterator<char>(input)), std::istreambuf_iterator<char>());
+      if (gz)
```

Comment:
the cool thing is, that it does gzip compression of the responses on the fly so that when graphreader comes at it with a `.gz` request it returns a gzipped payload with proper content encoding. otherwise it returns the raw bytes from disk. this lets us test all the options below